hsmd/hsmd.c

```diff
@@ -554,15 +598,66 @@ static void maybe_create_new_hsm(void)
-static void load_hsm(void)
+static void load_hsm(const struct secret *encryption_key)
 {
+	struct stat st;
 	int fd = open("hsm_secret", O_RDONLY);
 	if (fd < 0)
 		status_failed(STATUS_FAIL_INTERNAL_ERROR,
 			      "opening: %s", strerror(errno));
-	if (!read_all(fd, &secretstuff.hsm_secret, sizeof(secretstuff.hsm_secret)))
+	if (stat("hsm_secret", &st) != 0)
 		status_failed(STATUS_FAIL_INTERNAL_ERROR,
-			      "reading: %s", strerror(errno));
+		              "stating: %s", strerror(errno));
+
+	/* If the seed is stored in clear. */
+	if (st.st_size == 32) {
+		if (!read_all(fd, &secretstuff.hsm_secret, sizeof(secretstuff.hsm_secret)))
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			              "reading: %s", strerror(errno));
+		/* If an encryption key was passed with a not yet encrypted hsm_secret,
+		 * remove the old one and create an encrypted one. */
+		if (encryption_key) {
+			if (close(fd) != 0)
+				status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				              "closing: %s", strerror(errno));
+			if (remove("hsm_secret") != 0)
+				status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				              "removing clear hsm_secret: %s", strerror(errno));
+			maybe_create_new_hsm(encryption_key, false);
+			fd = open("hsm_secret", O_RDONLY);
+			if (fd < 0)
+				status_failed(STATUS_FAIL_INTERNAL_ERROR,
+				              "opening: %s", strerror(errno));
+		}
+	}
+	/*~ If an encryption key was passed and the `hsm_secret` is stored
+	 * encrypted, recover the seed from the cipher. */
+	if (encryption_key && st.st_size > 32) {
+		crypto_secretstream_xchacha20poly1305_state crypto_state;
+		u8 header[crypto_secretstream_xchacha20poly1305_HEADERBYTES];
+		/* The cipher size is static with xchacha20poly1305 */
+		u8 cipher[sizeof(struct secret) + crypto_secretstream_xchacha20poly1305_ABYTES];
+
+		if (!read_all(fd, &header, crypto_secretstream_xchacha20poly1305_HEADERBYTES))
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			              "Reading xchacha20 header: %s", strerror(errno));
+		if (!read_all(fd, cipher, sizeof(cipher)))
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			              "Reading encrypted secret: %s", strerror(errno));
+		if (crypto_secretstream_xchacha20poly1305_init_pull(&crypto_state, header,
+		                                                    encryption_key->data) != 0)
+			status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			              "Initializing the crypto state: %s", strerror(errno));
+		if (crypto_secretstream_xchacha20poly1305_pull(&crypto_state,
+		                                               secretstuff.hsm_secret.data,
+		                                               NULL, 0, cipher, sizeof(cipher),
+		                                               NULL, 0) != 0) {
+			/* Exit but don't throw a backtrace when the user made a mistake in typing
+			 * its password. Instead exit and `lightningd` will be able to give
+			 * an error message. */
+			exit(1);
+		}
+	}
```

Comment:
... else?

If it's encrypted, and they don't provide the key?  [Edit: ah, fixed in next commit!] 

Or if it's not, and they do?  Or if it's been truncated to 31 bytes?

Reply:
> Or if it's not, and they do?

Then we load the seed, delete the `hsm_secret` and call `create_encrypted_hsm_secret()`.

> Or if it's been truncated to 31 bytes?

Could it be ? Anyway the encrypted seed is always > 32 so I can make the condition a `<=`

Reply:
> If it's encrypted, and they don't provide the key? [Edit: ah, fixed in next commit!]

As a side not I handle this in `hsm_control` because I was not able to get `satus_failed` (with a new failure type) to send an error before replying to `init`. The msg type would always be zero. Hence the hacky "wrong password" case handling.